pkg/storage/chunk/prometheus_chunk.go

```diff
@@ -106,15 +162,43 @@ func (p *prometheusChunkIterator) Value() model.SamplePair {
 	}
 }
 
-func (p *prometheusChunkIterator) Batch(size int) Batch {
+func (p *prometheusChunkIterator) Batch(size int, valueType chunkenc.ValueType) Batch {
 	var batch Batch
+	var populate func(j int)
+	switch valueType {
+	case chunkenc.ValFloat:
+		batch.ValueTypes = chunkenc.ValFloat
+		batch.SampleValues = &[BatchSize]float64{}
+		populate = func(j int) {
+			t, v := p.it.At()
+			batch.Timestamps[j] = t
+			batch.SampleValues[j] = v
+		}
+	case chunkenc.ValHistogram:
+		batch.ValueTypes = chunkenc.ValHistogram
+		batch.HistogramValues = &[BatchSize]*histogram.Histogram{}
+		populate = func(j int) {
+			t, h := p.it.AtHistogram()
+			batch.Timestamps[j] = t
+			batch.HistogramValues[j] = h
+		}
+	case chunkenc.ValFloatHistogram:
+		batch.ValueTypes = chunkenc.ValFloatHistogram
+		batch.FloatHistogramValues = &[BatchSize]*histogram.FloatHistogram{}
+		populate = func(j int) {
+			t, fh := p.it.AtFloatHistogram()
+			batch.Timestamps[j] = t
+			batch.FloatHistogramValues[j] = fh
+		}
+	default:
+		panic(fmt.Sprintf("invalid chunk encoding %v", valueType))
+	}
+
 	j := 0
 	for j < size {
-		t, v := p.it.At()
-		batch.Timestamps[j] = t
-		batch.Values[j] = v
+		populate(j)
 		j++
-		if j < size && !p.it.Next() {
+		if j < size && p.it.Next() == chunkenc.ValNone {
```

Comment:
The `Batch` assumes all samples have the same type. We're not enforcing it here. As a safety net, I would check the type returned by `p.it.Next()`, compare it with the expected type (`valueType`) and if it's different than this should fail.

Reply:
fixed